### PR TITLE
AMReX_Config.H

### PR DESCRIPTION
**Pull request by WeiqunZhang**
Based on discussions with @sayerhs, @ax3l, and @mic84, We have decided to
make the following changes to the build systems.

All headers in Src/ will include AMReX_Config.H, which will be generated by
the build systems (both gnu make and cmake).  The file will include all the
macros we currently pass as compiler options.

The purpose of this is to solve some inconsistency issues in the current
systems.  We support both super build and using AMReX as a library.
Moreover, we do not want to force libamrex users to use the same build tool
that is used to build the library.  In this PR, we are put all the macro
definitions in AMReX_Config.H that is included by every AMReX header.

- [x] Insert AMReX_Config.H to all headers
- [x] Modify GNU make system (including the `configure` script) and make sure repeated `make`s do not trigger rebuild
- [x] CMake: no public defines anymore (already in find-package, now also in superbuild)
- [x] CMake: double-check that a 2nd call to make or make install does not rebuild everything (i.e. Config file generation as expected)
- [ ] CI: `cat` the generated `AMReX_Config.H` file
- [x] AmrVis
- [ ] AMR-Wind
- [x] Castro
- [x] IAMR
- [x] MAESTROeX
- [x] FHDeX
- [x] MFIX-Exa/CMake
- [x] MFIX-Exa/GNUMake
- [x] incflo
- [x] Nyx
- [ ] PeleC/CMake
- [x] PeleC/GNUMake
- [x] PeleLM
- [x] SedonaBox
- [ ] WarpX/CMake
- [x] WarpX/GNUMake

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
This might break application codes because we no longer pass `-D`s to the C++ compiler.  For example the code below might fail.
```
#ifndef MY_HEADER_H_
#define MY_HEADER_H_

#ifdef AMREX_PARTICLES
#include <AMReX_AmrParticles.H>
#endif

#include <AMReX_AmrLevel.H>
// .....

#endif 
```

There are two ways to fix this.  Move the `ifdef AMREX_PARTICLES` block after `#include <AMReX_AmrLevel.H>` or add `#include <AMReX_Config.H>` as the first header included.

File: Tools/CMake/AMReX_Config.H.in [hunk 1]
```diff
@@ -48,4 +49,7 @@
 #ifdef __cplusplus@COMP_DECLS@
 @OMP_DECLS@
 #endif
+#if defined(AMREX_USE_OMP) && !defined(_OPENMP)
+#error "libamrex was built with OpenMP"
```
Comment (ax3l) [inline Tools/CMake/AMReX_Config.H.in]:
```suggestion
#error "libamrex was built with OpenMP, so the downstream project must activate it, too"
```

**Review by ax3l — APPROVED** [review]
LGTM and compiles with the latest WarpX `development` branch as of 7fe38c8d13d9a511ab8f2c1243a4c58ca393c813

Tested for OpenMP superbuild:
```
cmake -S . -B build -DWarpX_amrex_repo=git@github.com:WeiqunZhang/amrex.git -DWarpX_amrex_branch=config_h
cmake --build build -j 2
```

**Comment by sayerhs** [comment]
@ax3l did we decide that we didn't need to `cat AMReX_Config.H` in CI because we are going to still pass this through `target_compile_definitions` for the `AMReX::amrex` target?

**Comment by ax3l** [comment]
@sayerhs I think we just forgot to add the `cat` in CI :thinking:

**Pull request merged**